## Introduction to Data Science

#### University of Redlands - DATA 101
#### Prof: Joanna Bieri [joanna_bieri@redlands.edu](mailto:joanna_bieri@redlands.edu)
#### [Class Website: data101.joannabieri.com](https://joannabieri.com/data101.html)

---------------------------------------
# Homework Day 2
---------------------------------------

GOALS:

1. Start using Python
2. Start looking at Data!

----------------------------------------------------------

This homework has **TWO** warm-up problems and **FOUR** problems.


## Hello World!

### Warm up Problem 1

Write python code that will print your name and run the cell.

In [2]:
print("Name: Elias Nicolas")

Name: Elias Nicolas


## Installing Modules

### Warm up Problem 2

Install the modules that you will need to run the rest of the code. You can copy and paste the commands from the class notes. This might take a minute or two to run... go get a cup of coffee :)

In [3]:
!conda install -y numpy
!conda install -y pandas
!conda install -y matplotlib
!conda install -y plotly
!conda install -y itables
!conda install -y statsmodels
!conda install -y -c conda-forge python-kaleido

Jupyter detected...
2 channel Terms of Service accepted
Channels:
 - conda-forge
 - defaults
Platform: osx-arm64
doneecting package metadata (repodata.json): - 
doneing environment: / 

# All requested packages already installed.

Jupyter detected...
2 channel Terms of Service accepted
Channels:
 - conda-forge
 - defaults
Platform: osx-arm64
doneecting package metadata (repodata.json): - 
doneing environment: / 

# All requested packages already installed.

Jupyter detected...
2 channel Terms of Service accepted
Channels:
 - conda-forge
 - defaults
Platform: osx-arm64
doneecting package metadata (repodata.json): - 
doneing environment: - 

# All requested packages already installed.

Jupyter detected...
2 channel Terms of Service accepted
Channels:
 - conda-forge
 - defaults
Platform: osx-arm64
doneecting package metadata (repodata.json): - 
doneing environment: / 

# All requested packages already installed.

Jupyter detected...
2 channel Terms of Service accepted
Channels:
 - conda-f

## Let's do some Data Science.

We will explore some data about how countries vote at the United Nations General Assembly. You should follow along with the notes and/or class video.

I am breaking this into parts and having you copy and paste the code so you can start to identify what the different parts of the code do.

-------------------------------
### Importing Packages:
-------------------------------

Every time we start a new project we will import the packages that will help us do the analysis. Copy and paste all of the imports in the cell below.

* numpy = mathematical and number packages for python
* pandas = pretty tables, dataframes, and data analysis packages
* matplotlib.pyplot = nice looking graphs
* plotly = nice looking graphs.
* itables = pretty looking tables that have a search bar

In [4]:
import pandas as pd 
import numpy as np 

import matplotlib.pyplot as plt
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.io as pio
pio.renderers.default='colab'

from itables import show

-------------------------------
### Get the Data:
-------------------------------

We will practice all sorts of ways to get data into our notebooks. When you use Pandas (pd) to read a csv:

    DF = pd.read_csv(file_location)

you are basically reading in a spreadsheet (Excel). The data for today is stored on my website.

It might take a minute for the data to load.

In [5]:
file_location = 'https://joannabieri.com/introdatascience/data/unvotes.csv'
DF = pd.read_csv(file_location)
years = [int(d.split('-')[0]) for d in DF['date']]
DF['year'] = years
DF = DF.drop('Unnamed: 0',axis=1)


-------------------------------
### Initial Data Exploration:
-------------------------------
-------------------------------
### Show the Data:
-------------------------------

We want to look at what we have. The data is now represented by the variable DF.

In [6]:
show(DF)

##prints number of rows
row_count = DF.shape[0]
print(row_count)

#prints number of columns 
column_count = DF.shape[1]
print(column_count)

#prints number of countries 
list_country = list(DF['country'].unique())
print(len(list_country))

857878
15
200


### Problem 1

Do some initial exploration of this data and write about it in the markdown cell below.

* How man columns are there?
* What do the columns represent?
* What countries are there?
* Can you search for a country you are interested in?
* How many rows are there?
* What other observations do you have?



**There are 15 columns.**

**The columns in this dataset represent the specifications of what different countries voted for on certain issues and when.**

**United States, Canada, Cuba, Dominican Republic, Mexico, Guatemala, Honduras, El Salvador, Nicaragua, Panama, Colombia, Venezuela, Ecuador, Peru, Brazil, Bolivia, Paraguay, Chile, Argentina, Uruguay, United Kingdom, Netherlands, Belgium, Luxembourg, France, Poland, Czechoslovakia, Yugoslavia, Greece, Russia, Ukraine, Belarus, Norway, Denmark, Liberia, Ethiopia, South Africa, Iran, Turkey, Iraq, Egypt, Syria, Lebanon, Saudi Arabia, Taiwan, India, Philippines, Australia, New Zealand, Costa Rica, Haiti, Sweden, Iceland, Afghanistan, Yemen Arab Republic, Pakistan, Thailand, Myanmar (Burma), Israel, Indonesia, Hungary, Jordan, Sri Lanka, Spain, Romania, Ireland, Portugal, Austria, Italy, Albania, Bulgaria, Finland, Morocco, Tunisia, Libya, Sudan, Nepal, Cambodia, Laos, Japan, Ghana, Malaysia, Guinea, Cyprus, Mali, Senegal, Benin, Niger, Côte d’Ivoire, Burkina Faso, Togo, Cameroon, Nigeria, Gabon, Central African Republic, Chad, Congo - Brazzaville, Somalia, Madagascar, Congo - Kinshasa, Mauritania, Sierra Leone, Mongolia, Tanzania, Jamaica, Trinidad & Tobago, Uganda, Burundi, Rwanda, Algeria, Kuwait, Kenya, Zanzibar, Malta, Zambia, Maldives, Malawi, Singapore, Guyana, Gambia, Lesotho, Barbados, Botswana, Yemen People's Republic, Mauritius, Equatorial Guinea, Eswatini, Fiji, Bahrain, Qatar, China, Oman, United Arab Emirates, Bhutan, Federal Republic of Germany, German Democratic Republic, Bahamas, Bangladesh, Grenada, Guinea-Bissau, Cape Verde, São Tomé & Príncipe, Mozambique, Comoros, Papua New Guinea, Suriname, Angola, Djibouti, Vietnam, Samoa, Seychelles, St. Lucia, Solomon Islands, Zimbabwe, Vanuatu, Antigua & Barbuda, Belize, St. Vincent & Grenadines, Dominica, St. Kitts & Nevis, Brunei, Liechtenstein, Namibia, Germany, Estonia, Latvia, Lithuania, Yemen, North Korea, South Korea, Marshall Islands, Micronesia (Federated States of), Croatia, Slovenia, Moldova, San Marino, Armenia, Azerbaijan, Bosnia & Herzegovina, Kazakhstan, Tajikistan, Turkmenistan, Kyrgyzstan, Andorra, Czechia, Slovakia, North Macedonia, Monaco, Georgia, Eritrea, Uzbekistan, Palau, Tonga, Nauru, Tuvalu, Kiribati, Switzerland, Timor-Leste, Montenegro, South Sudan**

**Yes, you can search for the county you are intrested in.**

**There are 857878 rows.** 

**N/A**

-------------------------------
### Make Python Explore the Data!:
-------------------------------

Copy and paste the command that will show you a list of the counties:

In [7]:
#prints the list of countries no quotes
list_country = list(DF['country'].unique())
print('[%s]'%', '.join(list_country))

##prints table
##show(pd.DataFrame(list_country, columns = ['country']))

##Shows data from a specific country
##DF[DF['country'] == 'Peru']

[United States, Canada, Cuba, Dominican Republic, Mexico, Guatemala, Honduras, El Salvador, Nicaragua, Panama, Colombia, Venezuela, Ecuador, Peru, Brazil, Bolivia, Paraguay, Chile, Argentina, Uruguay, United Kingdom, Netherlands, Belgium, Luxembourg, France, Poland, Czechoslovakia, Yugoslavia, Greece, Russia, Ukraine, Belarus, Norway, Denmark, Liberia, Ethiopia, South Africa, Iran, Turkey, Iraq, Egypt, Syria, Lebanon, Saudi Arabia, Taiwan, India, Philippines, Australia, New Zealand, Costa Rica, Haiti, Sweden, Iceland, Afghanistan, Yemen Arab Republic, Pakistan, Thailand, Myanmar (Burma), Israel, Indonesia, Hungary, Jordan, Sri Lanka, Spain, Romania, Ireland, Portugal, Austria, Italy, Albania, Bulgaria, Finland, Morocco, Tunisia, Libya, Sudan, Nepal, Cambodia, Laos, Japan, Ghana, Malaysia, Guinea, Cyprus, Mali, Senegal, Benin, Niger, Côte d’Ivoire, Burkina Faso, Togo, Cameroon, Nigeria, Gabon, Central African Republic, Chad, Congo - Brazzaville, Somalia, Madagascar, Congo - Kinshasa, Ma

Copy and paste the command that will count up the number of countries:

In [8]:
#prints number of countries 
list_country = list(DF['country'].unique())
print(f"There are {len(list_country)} different countries in the dataset")

There are 200 different countries in the dataset


### Problem 2

Try writing some code for your self. Above we found a countries list by focusing on the column named 'country'. See if you can redo that same code but change it to focus on the column named 'issue'.

What do I expect here:

* First copy and past the code from above
* Then change that code slightly
* Run the cell to see if it worked

In [9]:
#prints the list of various issues in a table 
issues_list = list(DF['issue'].unique())

show(pd.DataFrame(issues_list, columns = ['issue']))

-------------------------------
### Data Visualization:
-------------------------------

Now we will select three countries that we are interested in and see how their votes have changed over time. Below you should see code that selects: Turkey, United States, and United Kingdom.

**IMPORTANT** These have to be spelled and capitalized exactly like they are in the data. Python is unforgiving of typos!

You can just run the cell below - assuming you have done all the parts above!

In [10]:
countries = ['Peru', 'Ecuador', 'Canada']
issues = list(DF['issue'].unique())
c_groups = DF.groupby(['country','issue'])
print(issues)

['Human rights', 'Economic development', 'Colonialism', 'Palestinian conflict', 'Arms control and disarmament', 'Nuclear weapons and nuclear material']


Now that we have our focus countries, we can make pretty pictures.

You can just run the cell below - assuming you have done all the parts above!

In [11]:
def make_plot(countries,issue):
    '''
    A Python function that takes in the list of countries and issues and makes
    a scatter plot of each issue with a trendline for each country.
    '''
    x_data = []
    y_data = []
    c_data = []
    for cntry in countries:
        my_group = c_groups.get_group((cntry,issue))
        for y in my_group['year'].unique():
            x_data.append(y)
            tot_yes = sum(my_group[my_group['year']==y]['vote']=='yes')
            percent_yes = tot_yes/len(my_group[my_group['year']==y])*100
            y_data.append(percent_yes)
            c_data.append(cntry)

    fig = px.scatter(x=x_data, y=y_data,color=c_data,trendline="lowess",labels={"color": "Country"})

    fig.update_layout(
        title={
            'text': issue + '<br>',
            'y':0.9,
            'x':0.5,
            'xanchor': 'center',
            'yanchor': 'top'})
    fig.update_yaxes(title_text="% Yes")
    fig.update_xaxes(title_text="Year")
    fig.show()
    
for iss in issues:
    make_plot(countries,iss)

### Problem 3

Choose one of the graphs produced above (should match the lecture notes) and discuss what you see there:

1. Which issue are you focusing on.
2. What do the graph axis represent?
3. Say in words how the United States votes have changed over time.
4. How do the United States votes compare to the other countries?
5. See if you can come up with one or two questions that interest you about the data in the graph.

**I am focusing on the issue Palenstinian Conflict, analzing the United States, United Kingdom, and Turkey specifically.**

**The x-axis represents the year the vote took place and the y-axis represents the percentage of countries that said yes on the issue at hand.**

**The United States have voted "No" more and more over time, seeing a decrease over time in the percentage of "Yes" votes.**

**United States are the polar opposite compared to other countries, with most countries voting "Yes" on the issue in more recent times while the United States tends to vote "No".**

**Question 1: What cause the United States and United Kingdom to essentially switch roles over time?**

**Question 2: What United States party had more influence in office when these decisions were made, does Republican or Democratic influence matter in these foreign issues?**

### Problem 4

Now go back up to where we picked our three focus countries and choose some different ones. Rerun the analysis. If you want you could choose more than three countries. 

Discuss the graph for the same issue but with the new countries.

**I am still discussing the same issues from the previous section, the Palenstinian Conflict. The countries I am focusing on are Canada, Peru, and Ecuador.**

**Canada has a very similar graph trajectory to the United States, with the percentages of votes in the "Yes" direction decreasingly drastically during recent times.**

**Peru and Ecuador on the other had are on the other end of the spectrum, where their percentage of "Yes" votes range from 80 to 100 percent.**

**Peru and Ecuador share a very similar voting pattern while Canada seems to vote similarly to the United States, does geography play a role in voting? If so then why are western countries voting "No" on this subject while South American countries are voting "Yes".**

--------------------------------
### You are done with the homework... now what?
--------------------------------

1. Save your changes.
2. In the Git tab, **Stage** your changed files - use the (+) button next to the file.
3. **Commit** your changes by entering a summary and pushing commit.
4. **Push** your changes using the cloud button.
5. Check that your changes are on your GitHub repo (online)
6. **TAKE THE DAILY QUIZ ON CANVAS** the quiz will ask you to copy and past the HTTPS link to your repo for the day.
7. Come to class and get your questions answered.

At the end of the week  you will submit your final versions of the homework on Canvas for grading. For this you can just drag and drop the HW.ipynb files into canvas. 

**IMPORTANT** If this is confusing this first week, don't worry, I can help you and I will be really flexible about these first few deadlines.


In [ ]:
print("Done")